### PR TITLE
lib/generators.nix: init toHyprconf generator

### PR DESCRIPTION
**Pull request by khaneliman**
### Description
Creating a library module to be used amongst the various hypr ecosystem modules. In response to https://github.com/nix-community/home-manager/pull/5324#issuecomment-2080203195. I had to refactor the function when moving to a separate module so I created a test to verify that it didn't break anything. 

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC
@fufexan

### CODE REVIEW

File: modules/lib/hyprlang.nix
Comment (rycee) [inline modules/lib/hyprlang.nix:37]: Maybe `importantFields` should be extracted into a function argument?

Comment (khaneliman) [reply modules/lib/hyprlang.nix:44]: @fufexan  should these be hardcoded in each module as an argument or would it be another config option from the module to pass to the library?

Comment (khaneliman) [reply modules/lib/hyprlang.nix]: Decided to make it an option... I could see a benefit to a user being able to organize their config themself through it.

File: modules/lib/hyprlang.nix
Comment (fufexan) [inline modules/lib/hyprlang.nix:9]: Maybe having an argset instead of multiple args would be better, given their number. Also allows us to set defaults.

File: modules/lib/hyprlang.nix
Comment (rycee) [inline modules/lib/hyprlang.nix:9]: It seems this contains quite a few things that are very specific to Hyprland. Consider, for example, a program other than Hyprland that uses "hyprlang" and a configuration entry "device"…

I would suggest, for example, to instead allow `toHyprconf` to either take an attribute set or a list of attribute sets as input. If the list method is chosen then the output is in that order. The Hyprland module would then be responsible for doing the device and important prefixes stuff and constructing the list accordingly. This function shouldn't have to know anything Hyprland specific.

Also, the function should be placed in `generators.nix` next to `toKDL` 🙂

Comment (khaneliman) [reply modules/lib/hyprlang.nix:9]: Hmm.. alright, I'll look into that... If we go this route of moving the generator out of the library module, does the valueType need to go back to being declared in each module.

Comment (fufexan) [reply modules/lib/hyprlang.nix:9]: Apparently so. I think a Nixpkgs `pkgs/pkgs-lib/formats.nix`-style file would fare better in this context.

To be honest, I still don't understand the separation in Nixpkgs between `formatters` and `generators`. But that is off-topic.

Comment (khaneliman) [reply modules/lib/hyprlang.nix:9]: In implementing this... is there any actual reason for the ordering of specific sections? ie: the device configuration in the hyprland example being at the top or was it simply preference of the name attribute at the top of the attribute set?

Comment (khaneliman) [reply modules/lib/hyprlang.nix]: Update includes this change where it can function handling the list of attribute sets properly without needing to manually specify them. Also added another test so we can see it properly handling a singular device attribute set and a list of device attribute sets.

Comment (fufexan) [reply modules/lib/hyprlang.nix]: > In implementing this... is there any actual reason for the ordering of specific sections? ie: the device configuration in the hyprland example being at the top or was it simply preference of the name attribute at the top of the attribute set? Because I can just set up an important prefix in modules for the attributes we want generated at the top of their sections, if that was the reason.

The device section can be anywhere. But the `name` field has to be the first in the section, that's how vaxry implemented the parsing.

**Comment by rycee** [comment]
Thanks, nice improvements! It all looks OK to me but I'll let @fufexan have a change to look. I'll merge tomorrow unless any complaints arise.

**Comment by khaneliman** [comment]
> Thanks, nice improvements! It all looks OK to me but I'll let @fufexan have a change to look. I'll merge tomorrow unless any complaints arise.

Sounds good, I have these others that will follow using this generator.

- https://github.com/nix-community/home-manager/pull/5324
- https://github.com/nix-community/home-manager/pull/5325
- https://github.com/nix-community/home-manager/pull/5344

File: modules/lib/generators.nix
Comment (fufexan) [inline modules/lib/generators.nix:4]: Imo lib functions should be inherited, as it's being done in the `toKDL` function, knowing `with x;` is an anti-pattern in the Nix language.

**Review by fufexan — APPROVED** [review]

**Comment by khaneliman** [comment]
> Thanks, good job! Merged to master now 🙂
> 
> Note, I made a slight change in the function definition to avoid the frequent list generation and concatenation.

Thanks, I can rebase the others when I get home.

**Comment by khaneliman** [comment]
@rycee @fufexan the other module PRs should be ready for review now that the generator is set up. Do you want me to keep them as separate PRs or merge them into one to handle the conflicts that will arise from adding multiple news items and the module imports?

**Comment by rycee** [comment]
Keep them separate. I can resolve the merge conflicts as they come.

**Pull request merged**
